src/backend/drm/atomic/session.rs

```diff
@@ -0,0 +1,99 @@
+//!
+//! Support to register an open [`AtomicDrmDevice`](AtomicDrmDevice)
+//! to an open [`Session`](::backend::session::Session).
+//!
+
+use drm::control::crtc;
+use drm::Device as BasicDevice;
+use nix::libc::dev_t;
+use nix::sys::stat;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::rc::{Rc, Weak};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use super::{AtomicDrmDevice, AtomicDrmSurfaceInternal, Dev};
+use crate::backend::session::{AsSessionObserver, SessionObserver};
+
+/// [`SessionObserver`](SessionObserver)
+/// linked to the [`AtomicDrmDevice`](AtomicDrmDevice)
+/// it was created from.
+pub struct AtomicDrmDeviceObserver<A: AsRawFd + 'static> {
+    dev: Weak<Dev<A>>,
+    dev_id: dev_t,
+    privileged: bool,
+    active: Arc<AtomicBool>,
+    backends: Weak<RefCell<HashMap<crtc::Handle, Weak<AtomicDrmSurfaceInternal<A>>>>>,
+    logger: ::slog::Logger,
+}
+
+impl<A: AsRawFd + 'static> AsSessionObserver<AtomicDrmDeviceObserver<A>> for AtomicDrmDevice<A> {
+    fn observer(&mut self) -> AtomicDrmDeviceObserver<A> {
+        AtomicDrmDeviceObserver {
+            dev: Rc::downgrade(&self.dev),
+            dev_id: self.dev_id,
+            active: self.active.clone(),
+            privileged: self.dev.privileged,
+            backends: Rc::downgrade(&self.backends),
+            logger: self.logger.clone(),
+        }
+    }
+}
+
+impl<A: AsRawFd + 'static> SessionObserver for AtomicDrmDeviceObserver<A> {
+    fn pause(&mut self, devnum: Option<(u32, u32)>) {
+        if let Some((major, minor)) = devnum {
+            if major as u64 != stat::major(self.dev_id) || minor as u64 != stat::minor(self.dev_id) {
+                return;
+            }
+        }
+
+        // TODO: Clear overlay planes (if we ever use them)
```

Comment:
Do I understand correctly that this TODO is destinated to our future selves when we'll implement overlay planes support?

Reply:
Yes. So I do not forget, where I needed to do that. :sweat_smile: